File: js/vocab.js
// Token vocabulary for the move-prediction model.
//
// The dictionary itself is STATIC: js/vocab-data.js (generated once by
// tools/build-vocab.mjs, checked into git) holds the frozen, ordered token
// list — index in that array is the model's token id. This module wraps it
// with lookups and the SAN <-> token-sequence converters.
//
// A move is a SEQUENCE of tokens: optional modifier tokens followed by one
// core move token. The core token is what terminates a move — when the model
// emits it, the move is over. Canonical order:
//
//   [#|+]  [x]  [=Q|=R|=B|=N]  [core]
//
//   fxe8=Q+  ->  ['+', 'x', '=Q', 'fe8']     (pawn capture-promotion, check)
//   Nxf3     ->  ['x', 'Nf3']
//   Raxe1#   ->  ['#', 'x', 'Rae1']
//   e4       ->  ['e4']
//
// The core keeps SAN's disambiguation, so tokenization is lossless: 'fe8' is
// the pawn-capture core (source file + destination), 'Rae1'/'R1e2'/'Qh4e1'
// keep their file/rank/square disambiguators, castling cores are 'O-O' and
// 'O-O-O'.
//
// The three "nerf" tokens sit at the very end of the vocabulary. They are
// only ever valid as the FIRST token of a move sequence: sampling one means
// "the coming move is an inaccuracy/mistake/blunder", and the engine then
// re-queries the model conditioned on that quality.
//
// When your real tokenizer is ready, swap vocab-data.js for its dictionary
// (and adjust sanToTokens/tokensToSan here if the scheme differs).

import { TOKENS, CORE_COUNT } from './vocab-data.js';

export { TOKENS };

export const MODIFIER_TOKENS = ['+', '#', 'x', '=Q', '=R', '=B', '=N'];
export const NERF_TOKENS = ['<inaccuracy>', '<mistake>', '<blunder>'];

// vocab v2 — structural framing tokens used by the training pipeline, appended
// after the nerf tokens (see tools/build-vocab.mjs). They never appear as legal
// moves, so the engine's legality mask zeroes them during play; they exist so
// the model can be trained/prompted with game boundaries and Elo conditioning.
// <bos>/<eos> wrap each game; <elo-*> buckets encode both players' strength.
export const BOS_TOKEN = '<bos>';
export const EOS_TOKEN = '<eos>';
export const ELO_TOKENS = [
  '<elo-u800>', '<elo-800>', '<elo-1000>', '<elo-1200>', '<elo-1400>', '<elo-1600>',
  '<elo-1800>', '<elo-2000>', '<elo-2200>', '<elo-2400>', '<elo-2600>', '<elo-2800>', '<elo-3000p>',
];
// "strength unspecified" sentinel — feed this (per player) when Elo is unknown or
// you want unconditioned play. Training swaps real buckets to it for a fraction of
// games so the model learns that mode too. (vocab v3 appended the end tokens
// after it — the invariant is pure append, never reorder.)
export const ELO_ANY_TOKEN = '<elo-any>';
export const STRUCTURAL_TOKENS = [BOS_TOKEN, EOS_TOKEN, ...ELO_TOKENS, ELO_ANY_TOKEN];

// vocab v3 — game-end tokens, appended after the structural block. The packer
// emits one just before <eos> naming HOW the game ended and WHO acted, so the
// model learns to SAY how games end. Resignations are colour-differentiated
// because they happen on either turn (players often resign right after their
// own move) — colour cannot be inferred from parity. Flags are colour-
// differentiated for uniformity (the flagging side is always the side to
// move). <draw> carries no colour: the outcome is symmetric.
//
// During play only two are ever sampleable, and only as the FIRST token of
// the computer's move: the engine's OWN resign token (it resigns) and <draw>
// (it offers you a draw; declining masks <draw> for the rest of the game).
// The opponent's resign token and both flag tokens are gauges only. None of
// them ever enter the fed-back history — the game is over, or the offer was
// declined and the token rejected.
export const WHITE_RESIGN_TOKEN = '<white-resign>';
export const BLACK_RESIGN_TOKEN = '<black-resign>';
export const WHITE_FLAG_TOKEN = '<white-flag>';
export const BLACK_FLAG_TOKEN = '<black-flag>';
export const DRAW_TOKEN = '<draw>';
export const END_TOKENS = [
  WHITE_RESIGN_TOKEN, BLACK_RESIGN_TOKEN, WHITE_FLAG_TOKEN, BLACK_FLAG_TOKEN, DRAW_TOKEN,
];

// The resign token belonging to the side to move ('w' | 'b').
export const resignTokenFor = (turn) => (turn === 'w' ? WHITE_RESIGN_TOKEN : BLACK_RESIGN_TOKEN);

// Map a numeric Elo to its bucket token. Must match elo_bucket_token() in
// chess-tokeniser/vocab.py so training data and the harness agree.
export function eloBucketToken(elo) {
  if (!Number.isFinite(elo) || elo < 800) return '<elo-u800>';
  if (elo >= 3000) return '<elo-3000p>';
  return `<elo-${Math.floor(elo / 200) * 200}>`;
}

// nerf token -> quality label passed back into model.predict({quality})
export const QUALITY_BY_TOKEN = {
  '<inaccuracy>': 'inaccuracy',
  '<mistake>': 'mistake',
  '<blunder>': 'blunder',
};

// quality label -> conventional annotation glyph, used in the move list
export const GLYPH_BY_QUALITY = {
  inaccuracy: '?!',
  mistake: '?',
  blunder: '??',
};

export const CORE_TOKENS = Object.freeze(TOKENS.slice(0, CORE_COUNT));
export const TOKEN_INDEX = new Map(TOKENS.map((t, i) => [t, i]));
export const VOCAB_SIZE = TOKENS.length;

// The dictionary must end with exactly the modifier + nerf + structural tokens
// this module (and the engine) reason about. Fail loudly if the data file has
// drifted.
{
  const tail = TOKENS.slice(CORE_COUNT);
  const expected = [...MODIFIER_TOKENS, ...NERF_TOKENS, ...STRUCTURAL_TOKENS, ...END_TOKENS];
  if (tail.length !== expected.length || tail.some((t, i) => t !== expected[i])) {
    throw new Error('vocab-data.js is out of sync with vocab.js — regenerate it: node tools/build-vocab.mjs');
  }
}

export const isNerfToken = (t) => Object.hasOwn(QUALITY_BY_TOKEN, t);
export const isEndToken = (t) => END_TOKENS.includes(t);

const SAN_RE = /^([KQRBN]?[a-h]?[1-8]?)(x?)([a-h][1-8])(=[QRBN])?([+#])?$/;
const CASTLE_RE = /^(O-O(?:-O)?)([+#])?$/;

// SAN string -> canonical token sequence. Throws on strings that are not SAN.
export function sanToTokens(san) {
  const tokens = [];
  const castle = san.match(CASTLE_RE);
  if (castle) {
    if (castle[2]) tokens.push(castle[2]);
    tokens.push(castle[1]);
    return tokens;
  }
  const m = san.match(SAN_RE);
  if (!m) throw new Error(`not a SAN move: "${san}"`);
  const [, head, capture, dest, promo, suffix] = m;
  if (suffix) tokens.push(suffix);
  if (capture) tokens.push('x');
  if (promo) tokens.push(promo);
  tokens.push(head + dest);
  return tokens;
}

// Token sequence -> SAN string (inverse of sanToTokens).
export function tokensToSan(tokens) {
  let suffix = '', capture = '', promo = '';
  let core = null;
  for (const t of tokens) {
    if (t === '+' || t === '#') suffix = t;
    else if (t === 'x') capture = 'x';
    else if (t.startsWith('=')) promo = t;
    else core = t;
  }
  if (!core) throw new Error(`token sequence has no core move: [${tokens.join(', ')}]`);
  if (core.startsWith('O-O')) return core + suffix;
  const head = core.slice(0, -2);
  const dest = core.slice(-2);
  return head + capture + dest + promo + suffix;
}

File: chess-tokeniser/vocab.py
"""The frozen token vocabulary, shared by the training pipeline and the harness.

`js/vocab-data.js` is the single source of truth for token <-> id: its array
index IS the model's token id. This module parses that file (rather than
re-deriving the dictionary in Python) so the tokeniser, the packed training
data, the trained model, and the JS harness can never disagree on ids.

It also owns the translation from the *bracketed* surface syntax the tokeniser
emits (`[Nf3]`, `[x]`, `[??]`) to the *bare* forms the dictionary stores
(`Nf3`, `x`, `<blunder>`), and the game-framing used to pack games into one
stream:

    <bos> <elo-W> <elo-B>  <move tokens...>  [<resign>|<draw>]  <eos>

The <bos>/<eos>/<elo-*> tokens are "vocab v2", the <resign>/<draw> game-end
tokens are "vocab v3" — both pure appends to js/vocab-data.js that disturb no
pre-existing id.
"""
from __future__ import annotations

import os
import re
from functools import lru_cache

# ---------------------------------------------------------------- locating the dictionary

_HERE = os.path.dirname(os.path.abspath(__file__))
DEFAULT_VOCAB_JS = os.path.normpath(os.path.join(_HERE, "..", "js", "vocab-data.js"))

# ---------------------------------------------------------------- surface translation

# The three accuracy glyphs the tokeniser emits map to the dictionary's nerf
# words. Everything else in the dictionary is stored bare, so translating a
# tokeniser token is just "strip the brackets, then remap glyphs".
GLYPH_TO_NERF = {"??": "<blunder>", "?": "<mistake>", "?!": "<inaccuracy>"}

BOS = "<bos>"
EOS = "<eos>"
ELO_ANY = "<elo-any>"        # "strength unspecified" sentinel (see encode_game)
# vocab v3 game-end tokens, emitted just before <eos> (see Vocab._end_id).
# Resigns/flags are colour-differentiated: resignations happen on either turn
# (often right after the loser's own move), so colour is not inferable from
# parity; flags get colour too for uniformity. <draw> is symmetric — no colour.
WHITE_RESIGN = "<white-resign>"
BLACK_RESIGN = "<black-resign>"
WHITE_FLAG = "<white-flag>"
BLACK_FLAG = "<black-flag>"
DRAW = "<draw>"


def translate(token: str) -> str:
    """Bracketed tokeniser token -> bare dictionary surface form.

    '[Nf3]' -> 'Nf3' ; '[x]' -> 'x' ; '[=Q]' -> '=Q' ; '[??]' -> '<blunder>'.
    Tolerant of already-bare input (returns it unchanged apart from glyphs).
    """
    inner = token[1:-1] if len(token) >= 2 and token[0] == "[" and token[-1] == "]" else token
    return GLYPH_TO_NERF.get(inner, inner)


def elo_bucket_token(elo: int | float | None) -> str:
    """Numeric Elo -> bucket token. 200-point steps, open-ended tails.

    Mirrors eloBucketToken() in js/vocab.js. Non-numeric / <800 -> '<elo-u800>'.
    """
    try:
        e = int(elo)
    except (TypeError, ValueError):
        return "<elo-u800>"
    if e < 800:
        return "<elo-u800>"
    if e >= 3000:
        return "<elo-3000p>"
    return f"<elo-{(e // 200) * 200}>"


# ---------------------------------------------------------------- the dictionary

_TOKENS_RE = re.compile(r"TOKENS\s*=\s*Object\.freeze\(\[(.*?)\]\)", re.S)
_QUOTED_RE = re.compile(r"'((?:[^'\\]|\\.)*)'")
_CORE_COUNT_RE = re.compile(r"CORE_COUNT\s*=\s*(\d+)")


class Vocab:
    """Ordered token list + id lookups, loaded from js/vocab-data.js."""

    def __init__(self, tokens: list[str], core_count: int):
        self.tokens = tokens
        self.core_count = core_count
        self.stoi = {t: i for i, t in enumerate(tokens)}
        self.itos = {i: t for i, t in enumerate(tokens)}
        self.size = len(tokens)
        # structural ids used during framing (fail loudly if vocab v2/v3 is missing)
        missing = [t for t in (BOS, EOS, ELO_ANY, "<elo-u800>", "<elo-3000p>",
                               WHITE_RESIGN, BLACK_RESIGN, WHITE_FLAG, BLACK_FLAG, DRAW)
                   if t not in self.stoi]
        if missing:
            raise ValueError(
                f"vocab-data.js is missing structural tokens {missing}; "
                "regenerate it with: node tools/build-vocab.mjs"
            )
        self.bos_id = self.stoi[BOS]
        self.eos_id = self.stoi[EOS]
        self.elo_any_id = self.stoi[ELO_ANY]
        self.white_resign_id = self.stoi[WHITE_RESIGN]
        self.black_resign_id = self.stoi[BLACK_RESIGN]
        self.white_flag_id = self.stoi[WHITE_FLAG]
        self.black_flag_id = self.stoi[BLACK_FLAG]
        self.draw_id = self.stoi[DRAW]

    # -- framing -------------------------------------------------------------

    def _elo_id(self, elo, elo_dropout, rng):
        """Bucket id for `elo`, or the <elo-any> sentinel with prob elo_dropout."""
        if elo_dropout and rng is not None and rng.random() < elo_dropout:
            return self.elo_any_id
        return self.stoi[elo_bucket_token(elo)]

    def _end_id(self, result, termination, saw_mate: bool):
        """Game-end token id for HOW the game ended, or None for a bare <eos>.

        Header-only classification (the pipeline never replays the board), with
        the actor's COLOUR taken from Result — never from parity, because
        players often resign right after their own move, i.e. on the winner's
        turn (flags, by the clock rules, always belong to the side to move, but
        the colour token keeps them uniform and parity-proof too):
          * <draw>  — Result 1/2-1/2 with Termination "Normal". Includes
            rule-draws (stalemate/repetition/...): those positions are terminal
            for the harness anyway, and "offers a draw in a dead position" is
            the behaviour we want. Timeout-draws (flag vs. insufficient
            material, Termination "Time forfeit") stay bare — a flag token
            implies a loss and <draw> would mislabel the clock death.
          * <white-resign>/<black-resign> — decisive Result with Termination
            "Normal" and no mate in the movetext (mate is its own signal); the
            token names the LOSER. Both parities are kept: at play the engine
            only ever samples its OWN colour's resign token on its own turn,
            so opposite-parity occurrences are gauge/context signal, never a
            "resign while winning" hazard.
          * <white-flag>/<black-flag> — decisive Result with Termination
            "Time forfeit"; the token names the LOSER. Never sampled in play
            (the harness has no clocks) — they exist so time-forfeit games
            don't end with an uninformative bare <eos> that teaches "games
            just stop after ordinary moves".
        Everything else (Abandoned, Rules infraction, '*', missing headers)
        keeps the bare <eos>.
        """
        if result == "1/2-1/2":
            return self.draw_id if termination == "Normal" else None
        if result not in ("1-0", "0-1"):
            return None
        white_lost = result == "0-1"
        if termination == "Normal" and not saw_mate:
            return self.white_resign_id if white_lost else self.black_resign_id
        if termination == "Time forfeit":
            return self.white_flag_id if white_lost else self.black_flag_id
        return None

    def encode_game(self, tokens_field: str, white_elo, black_elo, elo_dropout: float = 0.0, rng=None,
                    result=None, termination=None):
        """Frame one game as ids: <bos> <elo-W> <elo-B> <moves...> [<end>] <eos>.

        `tokens_field` is the space-joined bracketed stream from build_dataset.
        With `elo_dropout` > 0 and a `random.Random` `rng`, each Elo slot is
        INDEPENDENTLY replaced by <elo-any> with that probability — so the model
        also sees unconditioned / half-conditioned games and inference can omit
        or neutralise Elo. rng makes the choice reproducible; pass a seeded one.

        `result` ("1-0"/"0-1"/"1/2-1/2") and `termination` (the Lichess
        Termination header) drive the colour-differentiated game-end token
        (see _end_id); left at None — e.g. when packing pre-v3 shards — the
        game ends with a bare <eos> exactly as before.

        Returns (ids, unknown_tokens). If any move token is outside the frozen
        dictionary, ids is None and unknown_tokens names the offenders (the
        caller should drop the game). On real Lichess data this never fires —
        the dictionary is a superset of legal SAN — but synthetic corpora with
        geometrically-loose SANs can trip it, which is the point of the check.
        """
        ids = [self.bos_id, self._elo_id(white_elo, elo_dropout, rng), self._elo_id(black_elo, elo_dropout, rng)]
        unknown: set[str] = set()
        saw_mate = False
        for tok in tokens_field.split():
            surface = translate(tok)
            sid = self.stoi.get(surface)
            if sid is None:
                unknown.add(tok)
                continue
            ids.append(sid)
            if surface == "#":
                saw_mate = True
        if unknown:
            return None, unknown
        end_id = self._end_id(result, termination, saw_mate)
        if end_id is not None:
            ids.append(end_id)
        ids.append(self.eos_id)
        return ids, unknown

    def decode(self, ids) -> list[str]:
        return [self.itos[i] for i in ids]


def load_vocab(path: str | None = None) -> Vocab:
    with open(path or DEFAULT_VOCAB_JS, encoding="utf-8") as f:
        src = f.read()
    m = _TOKENS_RE.search(src)
    if not m:
        raise ValueError(f"could not find TOKENS array in {path or DEFAULT_VOCAB_JS}")
    tokens = [t.replace("\\'", "'") for t in _QUOTED_RE.findall(m.group(1))]
    if not tokens:
        raise ValueError("parsed zero tokens from vocab-data.js")
    cc = _CORE_COUNT_RE.search(src)
    return Vocab(tokens, int(cc.group(1)) if cc else len(tokens))


@lru_cache(maxsize=None)
def get_vocab(path: str | None = None) -> Vocab:
    """Cached loader for the default (or a specific) dictionary file."""
    return load_vocab(path)


if __name__ == "__main__":
    v = get_vocab()
    print(f"{v.size} tokens ({v.core_count} cores) from {DEFAULT_VOCAB_JS}")
    print(f"bos={v.bos_id} eos={v.eos_id}")
    ids, unk = v.encode_game("[Nf3] [x] [Bb5] [??] [e5] [+] [Qf5]", 1834, 1790)
    print("sample ids:", ids)
    print("decoded:", v.decode(ids))
